src/librustc_mir/transform/qualify_consts.rs

```diff
@@ -1038,32 +1038,28 @@ impl<'a, 'tcx> Checker<'a, 'tcx> {
 
         debug!("qualify_const: promotion_candidates={:?}", self.promotion_candidates);
         for candidate in &self.promotion_candidates {
-            match *candidate {
+            let promoted_place = match *candidate {
                 Candidate::Repeat(Location { block: bb, statement_index: stmt_idx }) => {
-                    if let StatementKind::Assign(box(_, Rvalue::Repeat(
-                        Operand::Move(Place {
-                            base: PlaceBase::Local(index),
-                            projection: box [],
-                        }),
-                        _
-                    ))) = self.body[bb].statements[stmt_idx].kind {
-                        promoted_temps.insert(index);
+                    match &self.body[bb].statements[stmt_idx].kind {
+                        StatementKind::Assign(box(_, Rvalue::Repeat(Operand::Move(place), _)))
+                            => place,
+                        _ => continue,
                     }
                 }
                 Candidate::Ref(Location { block: bb, statement_index: stmt_idx }) => {
-                    if let StatementKind::Assign(
-                        box(
-                            _,
-                            Rvalue::Ref(_, _, Place {
-                                base: PlaceBase::Local(index),
-                                projection: box [],
-                            })
-                        )
-                    ) = self.body[bb].statements[stmt_idx].kind {
-                        promoted_temps.insert(index);
+                    match &self.body[bb].statements[stmt_idx].kind {
+                        StatementKind::Assign(box( _, Rvalue::Ref(_, _, place))) => place,
+                        _ => continue,
                     }
                 }
-                Candidate::Argument { .. } => {}
+                Candidate::Argument { .. } => continue,
+            };
+
+            match promoted_place.base {
+                PlaceBase::Local(local) if !promoted_place.is_indirect() => {
```

Comment:
The "indirect" thing is irrelevant AFAIK, all you care about is the base.

Reply:
I guess we never promote the result of a deref projection, so that guard will never take effect. I could make this a `bug` or just remove it.

Reply:
I've made this an `assert` for now so it will fail loudly on the test suite if my assumptions are wrong. Probably want to make this a `debug_assert` or remove it entirely before merging.

Reply:
You can just remove it IMO. We *really* need to stop using "delete `StorageDead`" for faux promotion, I dislike it more by the minute.

Reply:
The assertion has been removed in the latest force push.